power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -273,41 +366,550 @@ inline auto rank_transformers(State const& state) -> RankedTransformerGroups {
     return rank_transformers(get_edge_weights(build_transformer_graph(state)));
 }
 
-template <typename StateCalculator, typename StateUpdater_, typename State_>
-    requires detail::steady_state_calculator_c<StateCalculator, State_> &&
-             std::invocable<std::remove_cvref_t<StateUpdater_>, ConstDataset const&>
-class TapPositionOptimizer : public detail::BaseOptimizer<StateCalculator, State_> {
+struct TransformerRanker {
+    template <main_core::main_model_state_c State>
+    auto operator()(State const& state) const -> RankedTransformerGroups {
+        return rank_transformers(state);
+    }
+};
+
+constexpr IntS tap_one_step_up(transformer_c auto const& transformer) {
+    auto const tap_pos = transformer.tap_pos();
+    auto const tap_max = transformer.tap_max();
+    return tap_pos < tap_max ? tap_pos + 1 : tap_max;
+}
+
+constexpr IntS tap_one_step_down(transformer_c auto const& transformer) {
+    auto const tap_pos = transformer.tap_pos();
+    auto const tap_min = transformer.tap_min();
+    return tap_pos > tap_min ? tap_pos - 1 : tap_min;
+}
```

Comment:
good question. i did want to add this to the tests (hence here the checkpoint) but tbh i don't know. there are 2 possible resolutions with opposite effect:

* [ ] `local_maximum` strategy results in the `tap_pos` that is closest to `127`.
  * this can be ensured by setting `tap_max = std::max(t.tap_min(), t.tap_max())` (idem for `tap_min`)
  * in that way, `tap_one_step_up` always amounts to `tap_pos = t.tap_pos() + 1` but clamped to the valid range
* [ ] `local_maximum` strategy returns the closest value to `t.tap_max()` (and vice versa for `local_minimum` strategy)
  * this can be achieved by setting `tap_pos -> t.tap_pos() + sgn(t.tap_max() - t.tap_min())`
  * in that way, the behaviour of `tap_one_step_up` depends on whether `t.tap_max() > t.tap_min()`

This requires expert input. @TonyXiang8787 do you know which of the two is desired?

Reply:
after discussion with tony, the answer is the 2nd option, which is the 3rd option + the assumption that the user correctly identifies `tap_max` as the tap position with the highest `u_rated` on the tap side, which is the lowest `u_rated` at the control side. This works because of the check that throws `AutomaticTapCalculationError` with the error message equivalent to `f"Automatic tap changing regulator with tap_side at LV side is not supported. Found at id {id}"`